src/python/pants/backend/python/tasks2/partition_targets.py

```diff
@@ -0,0 +1,146 @@
+# coding=utf-8
+# Copyright 2017 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import (absolute_import, division, generators, nested_scopes, print_function,
+                        unicode_literals, with_statement)
+
+from collections import namedtuple
+
+from pants.build_graph.target import Target
+from pants.task.task import Task
+
+
+class TargetsPartition(object):
+  """Represents a partitioning of a set of targets into non-empty subsets."""
+  def __init__(self, subsets):
+    """Constructs a new TargetsPartition from a given collection of subsets.
+
+    :param subsets: an iterable of iterables of targets.
+    """
+    if any(len(subset) == 0 for subset in subsets):
+      raise ValueError('Subsets must be non-empty.')
+    self._subsets = frozenset(frozenset(subset) for subset in subsets)
+    if self._subsets and (
+        len(frozenset.union(*self._subsets)) != sum(len(subset) for subset in self._subsets)):
+      raise ValueError('Subsets must be disjoint.')
+    self._subset_by_target = {target: subset for subset in self._subsets for target in subset}
+
+  @property
+  def subsets(self):
+    return self._subsets
+
+  def find_subset_for_target(self, target):
+    """Returns a subset containing the given targets.
+
+    Raises a KeyError if such target does not exist.
+    """
+    return self._subset_by_target[target]
+
+  def find_subset_for_targets(self, targets):
+    """Returns a single subset that contains all given targets.
+
+    If such subset does not exists within the partition a ValueError is raised.
+    """
+    if not self._subsets:
+      raise ValueError('No subsets.')
+    if not targets:
+      # Any subset will do.
+      return next(iter(self._subsets))
+    subset = self._subset_by_target.get(next(iter(targets)))
+    if not subset or not subset.issuperset(targets):
+      raise ValueError('No subset contains all targets: {}'.format(
+          ' '.join(tgt.address.spec for tgt in targets)))
+    return subset
+
+
+class PartitionTargets(Task):
+  """Generates partitions of the target roots.
+
+  This tasks produces `TargetPartition`s. Each representing a partition of the target roots into
+  non-empty disjoint subsets of targets based on a requested partitioning strategy.
+
+  There are several supported strategies to produce a TargetPartition from a set of target roots:
+
+  STRATEGY_MINIMAL:
+      partitions the target roots into the smallest possible number of subsets such that if one target
+      root depends on another, then both of them will be in the subset. This ensures
+      that incompatible target roots will fall into different subsets while minimizing the number
+      of chroots.
+  STRATEGY_PER_TARGET:
+      A partition in which Each target root will be in its own isolated subset.
```

Comment:
But the partitioning code is agnostic to what the underlying target set is, right? It could be target roots or it could be all the targets.  

Let me see if I can clarify my thinking, at least to myself, with an example of three target roots that depend on the same inner target:

R1-> T
R2 -> R3 -> T

The partitioning task produces one or more of the following 4 partitions:

global: (R1, R2, R3)
minimal: (R1)(R2)
per target root: (R1)(R3)(R3)
per target: (R1)(R2)(R3)(T)

Then, the interpreter selection task will select for (the transitive closure of) every subset of every partition.  But it doesn't need to know or care what's in those partitions. Only the consuming task (PytestRun, PythonEval, etc.) cares.

Does this make sense, or am I still missing something?